### PR TITLE
[mlir][LLVM] Verify too many indices in GEP verifier

### LINKED ISSUE #70168: [MLIR][LLVM] Add verifiers to the LLVM dialect GEP operation
Me and others have stumbled over invalid GEP operations due to inconsistencies between element types and the indices uses to index into the pointer. LLVM IR has strict verifiers for this, while the LLVM dialect does not. This can lead to unexpected crashes after exporting LLVM dialect to LLVM IR.

Given that GEP is considered the most complicated instruction in LLVM IR, it would be sensible to introduce verification for the GEP's indices in MLIR.

### PR DESCRIPTION
**Pull request by zero9178**
The current verifier stopped verification with a success value as soon as a type was encountered that cannot be indexed into. The correct behaviour in this case is to error out as there are too many indices for the element type. Not doing so leads to bad user-experience as an invalid GEP is likely to fail only later during LLVM IR translation.

This PR implements the correct verification behaviour. Some tests upstream had to also be fixed as they were creating invalid GEPs.

Fixes https://github.com/llvm/llvm-project/issues/70168

### CODE REVIEW

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 1]
```diff
@@ -664,90 +664,51 @@ static void printGEPIndices(OpAsmPrinter &printer, LLVM::GEPOp gepOp,
...
+                    GEPIndicesAdaptor<ValueRange> indices,
+                    function_ref<InFlightDiagnostic()> emitOpError) {
   if (indexPos >= indices.size())
     // Stop searching
-    return llvm::Error::success();
+    return success();
 
-  return llvm::TypeSwitch<Type, llvm::Error>(baseGEPType)
-      .Case<LLVMStructType>([&](LLVMStructType structType) -> llvm::Error {
+  return llvm::TypeSwitch<Type, LogicalResult>(baseGEPType)
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:677]:
```suggestion
  return TypeSwitch<Type, LogicalResult>(baseGEPType)
```
I believe the namespace is not needed here.

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
Please wait for the CI to pass and monitor the bots after your land it, in particular https://lab.llvm.org/buildbot/#/builders/61 which exercise integrations tests you may not run locally.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 2]
```diff
@@ -664,90 +664,51 @@ static void printGEPIndices(OpAsmPrinter &printer, LLVM::GEPOp gepOp,
...
-char GEPIndexError::ID = 0;
-char GEPIndexOutOfBoundError::ID = 0;
-char GEPStaticIndexError::ID = 0;
-
-/// For the given `structIndices` and `indices`, check if they're complied
-/// with `baseGEPType`, especially check against LLVMStructTypes nested within.
-static llvm::Error verifyStructIndices(Type baseGEPType, unsigned indexPos,
-                                       GEPIndicesAdaptor<ValueRange> indices) {
+/// For the given `indices`, check if they comply with `baseGEPType`,
+// especially check against LLVMStructTypes nested within.
```
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:668]:
```suggestion
/// especially check against LLVMStructTypes nested within.
```

**Review by gysit — APPROVED** [review]
Thanks for the quick fix!

LGTM

**Review by Dinistro — APPROVED** [review]
LGTM modulo nit. Thanks for the quick fix 😄

**Pull request merged**